src/tools/db-upgrade/DbUpgrader.cpp

```diff
@@ -691,11 +690,170 @@ void UpgraderSpace::encodeVertexValue(PartitionID partId,
+                case meta::cpp2::PropertyType::BOOL: {
+                    try {
+                        fval = val.getFloat();
+                        bval = folly::to<bool>(fval);
+                        val.setBool(bval);
+                        return WriteResult::SUCCEEDED;
+                    } catch (const std::exception& e) {
+                        return WriteResult::TYPE_MISMATCH;
+                    }
+                }
+                // other not need convert
+                default:
+                    return WriteResult::SUCCEEDED;
+            }
+        }
+        case Value::Type::STRING: {
+            switch (newpropType) {
+                case meta::cpp2::PropertyType::INT8:
+                case meta::cpp2::PropertyType::INT16:
+                case meta::cpp2::PropertyType::INT32:
+                case meta::cpp2::PropertyType::INT64:
+                case meta::cpp2::PropertyType::TIMESTAMP:
+                case meta::cpp2::PropertyType::VID: {
+                    try {
+                        sval = val.getStr();
+                        ival = folly::to<int64_t>(sval);
+                        val.setInt(ival);
+                        return WriteResult::SUCCEEDED;
+                    } catch (const std::exception& e) {
+                        return WriteResult::TYPE_MISMATCH;
+                    }
+                }
+                case meta::cpp2::PropertyType::BOOL: {
+                    try {
+                        sval = val.getStr();
+                        bval = folly::to<bool>(sval);
+                        val.setBool(bval);
+                        return WriteResult::SUCCEEDED;
+                    } catch (const std::exception& e) {
+                        return WriteResult::TYPE_MISMATCH;
+                    }
+                }
+                case meta::cpp2::PropertyType::FLOAT:
+                case meta::cpp2::PropertyType::DOUBLE: {
+                    try {
+                        sval = val.getStr();
+                        fval = folly::to<double>(sval);
+                        val.setFloat(fval);
+                        return WriteResult::SUCCEEDED;
+                    } catch (const std::exception& e) {
+                        return WriteResult::TYPE_MISMATCH;
+                    }
+                }
+                // other not need convert
+                default:
+                    return WriteResult::SUCCEEDED;
+            }
+        }
+        // other not need convert
+        default:
+            return WriteResult::SUCCEEDED;
+    }
+}
+
+
 // Used for vertex and edge
 std::string UpgraderSpace::encodeRowVal(const RowReader* reader,
                                         const meta::NebulaSchemaProvider* schema,
                                         std::vector<std::string>& fieldName) {
-    auto oldSchema = reader->getSchema();
+    auto oldSchema = dynamic_cast<const meta::NebulaSchemaProvider*>(reader->getSchema());
```

Comment:
No need to `dynamic_cast` here.

Reply:
Yes, I changed the convertValue parameter to SchemaProviderIf*